### PR TITLE
sutible for running on various clusters?

### PR DESCRIPTION
**Pull request by cmeesters**
Dear Paddy,

Meanwhile I found two use cases:
- simple programs, with +/- uniform run time (original case)
- the STAR aligner, where instances of the program can be distributed across several nodes. It scales well ... up to a memory block (NUMA) and the run times per instance do not vary much (provided the input is of similar size). And it is able to use shared memory: The first instance loads a reference into shared memory all other instances can access - provided that there is a *first* instance. This is what the --delay-option is for in my version.

Basically, there are a few things different to your original version:
- an easybuild config file (which still points to my fork)
- a cluster specific configuration file: I figure, what is different from cluster to cluster is the node local scratch directory. It has to be adapted manually.
- the command line parsing (I am using the argbash template, see https://argbash.io/ )
- some sanity checks
- difference between multithreaded applications and single threaded applications
- a --parameters-flag, it skips the tests for input files
- a dummy command is inserted, if the number of inputs is smaller than the number of job steps (which is acceptable in some cases).

If interested you should definitively adapt the credit section, because I omitted you. Also, the documentation HERE-doc needs revision. More recent documentation for our cluster can be found at:
https://mogonwiki.zdv.uni-mainz.de/dokuwiki/node_local_scheduling?&#the_zdv-taskfarm_script (and downwards).

Best regards,
Christian

### CODE REVIEW

**Comment by paddydoyle** [comment]
Hi Christian,

Thanks for the interest in the task farm script and for the pull request. I'm sorry it's taken me so long to replying; it caught me at a very busy time and I wanted to look at it properly before replying.

I'm afraid your code as it stands breaks backwards-compatibility with ours: the original `staskfarm commands.txt` method of running doesn't work. So I can't accept the PR as it stands because it would break the script for existing users here.

I do have some feedback if that's ok:

- a couple of typos 'SLRUM...'
- the final 'else' branch (our original `staskfarm commands.txt` way of running) is broken; the MULTIFILE does not exist in that branch of the script, and even if it does the staskfarm doesn't write the correct values to it, so it's not backwards-compatible with our version.
- it expects the SCRATCHDIR (default `/tmp/${SLURM_JOB_ID}`) to be pre-created on all the job nodes; that's not the case in our clusters.. do you have a local slurm plugin which does that? In our case, the SCRATCHDIR would have to be manually created by the sbatch script, which isn't something we want to enforce on the users at the moment.
- in particular actually, if SCRATCHDIR does not exist (e.g. when checking the help message on the headnode), the script won't even print the help/usage message; moving that test down in the code would help.
- I would suggest using `/bin/true` as your dummy command, rather than creating a no-op dummy.sh which must be sbcast'ed to all the nodes

And one big bit of feedback is regarding how it creates many `MULTIFILE` files. I believe that is not efficient. For example, if you have a 4-core job and 200 tasks, it will do the following:

- take the first 4 tasks, and create a 4-line multi-conf file.
- it will execute those tasks via srun --multi-prog, and wait for all 4 tasks to finish before proceeding. In particular, 3 of the 4 cores could be idle for a period of time.
- then it will take the next 4 tasks and create another 4-line mulit-conf file.
- it will execute those tasks via srun --multi-prog, and wait for all 4 tasks to finish before proceeding. In particular, 3 of the 4 cores could be idle for a period of time.
- and so on.

This has the effect that each time through, the execution time is bounded by the longest task in each batch of 4.

If you compare how we did it in the original staskfarm:

- create 4 script files, each containing 50 commands (it distributes them in a naive round-robin)
- create a single multi-conf file, pointing to those 4 scripts
- execute the 4 scripts via srun --multi-prog
- on each of the 4 processors, when the first task in the corresponding script finishes, it immediately moves onto the second, and so on

The execution time is now bounded by the longest set of 50 tasks, which is now more of a global bound, than for each batch of 4. There could still be 3 out of 4 processors idle at the end, but I believe in general it will lead to less idle time overall (at least from my testing it shows that).

Plus there is less interaction with the slurm daemon: 4 launches compared to 50. (Maybe that's not a big deal.)

In summary, your approach will work reasonably well when all tasks take roughly the same amount of time (likewise for mine). But if there is a variation in execution times, then my approach would tend to do a better job of hiding that variation.

Having said all of that, I do appreciate your use case of `cmd lots of parameters`; that's a nice addition and I can definitely see how it would be useful for people.

Thanks,
Paddy
